Segfault on a truncated PNG

```
loadImage segfaults on a truncated PNG instead of rejecting
```

**Body:**

## Summary

A PNG truncated mid-`IDAT` crashes the process with SIGSEGV rather than rejecting the promise. There is no error to catch, so a server handling untrusted or unreliable image input dies outright.

Related to #832 and #1255 in spirit — uncatchable failures on malformed input — but this is a raster path, not SVG, and it segfaults rather than hanging.

## Environment

| | |
|---|---|
| `@napi-rs/canvas` | 1.0.3 |
| binary | `skia.linux-x64-gnu.node` |
| Node | v22.22.2 |
| platform | linux-x64 |

## Reproduction

```js
import { createCanvas, loadImage } from '@napi-rs/canvas'

const canvas = createCanvas(64, 64)
const ctx = canvas.getContext('2d')
ctx.fillStyle = '#345678'
ctx.fillRect(0, 0, 64, 64)
const png = canvas.toBuffer('image/png')   // 239 bytes

const truncated = png.subarray(0, 40)      // cut mid-IDAT

console.log(`${truncated.length} bytes`)
try {
  const img = await loadImage(truncated)
  console.log('  ok', img.width, img.height)
} catch (e) {
  console.log('  threw:', e.message)
}
console.log('  reached end of script')
```

### Actual output

```
40 bytes
```

Exit code 139 (SIGSEGV). Neither the `catch` nor the final line is reached.

### Expected output

```
40 bytes
  threw: <some error>
  reached end of script
```

## Other inputs that crash the same way

All built from a valid PNG, all exit 139:

- PNG signature followed by random bytes and a well-formed `IEND`
- PNG truncated immediately after `IHDR`
- `IHDR` declaring colour type `9`
- `IHDR` declaring `0x0` dimensions
- a valid PNG whose first four bytes are replaced with the JPEG magic

One related input aborts rather than segfaulting — `IHDR` declaring `100000x100000` reaches a Skia assertion:

```
[skia] ** ERROR ** ../../src/core/SkBitmap.cpp:252: fatal error:
"assertf(this->tryAllocPixels(info, rowBytes)): ColorType:6 AlphaType:1 [w:100000 h:100000] rb:0"
```

(exit 132). That one at least prints something, but it is still not catchable.

Note that Skia is lenient with *some* damage — truncating the same 239-byte PNG to 119 bytes decodes fine, as do a corrupted `IDAT` payload and a bad `IDAT` CRC. It is specific shapes of malformation that crash.

## Why it matters

Any service that decodes images it did not produce — uploads, remote fetches, upstream APIs that can return a partial response — can be killed by a single malformed input. In a serverless function this is a 502 with no error body and nothing in the logs.

Rejecting the promise, as the empty-input path already does since #1255, would be enough.

--- a/__test__/loadimage.spec.ts
+++ b/__test__/loadimage.spec.ts
@@ -5,6 +5,7 @@ import { readFile } from 'node:fs/promises'
 import { URL, pathToFileURL, fileURLToPath } from 'node:url'
 
 import test from 'ava'
+import { crc32 } from '@node-rs/crc32'
 
 import { createCanvas, Image, loadImage } from '../index'
 
@@ -136,17 +137,6 @@ test('loadImage settles on invalid base64 data URL (issue #1255)', async (t) => 
 // Regression tests for https://github.com/Brooooooklyn/canvas/issues/1309
 // Run malformed inputs in child processes so a native crash is reported as a
 // test failure instead of terminating the entire test runner.
-
-function crc32(buffer: Buffer): number {
-  let crc = -1
-  for (const byte of buffer) {
-    crc ^= byte
-    for (let bit = 0; bit < 8; bit++) {
-      crc = (crc >>> 1) ^ (crc & 1 ? 0xedb88320 : 0)
-    }
-  }
-  return (crc ^ -1) >>> 0
-}
 
 function patchIhdr(png: Buffer, width: number, height: number, colorType?: number): Buffer {
   const result = Buffer.from(png)

--- a/__test__/loadimage.spec.ts
+++ b/__test__/loadimage.spec.ts
@@ -1,4 +1,4 @@
-import { spawnSync } from 'node:child_process'
+import { spawn } from 'node:child_process'
 import { join, dirname } from 'node:path'
 import fs from 'node:fs'
 import { readFile } from 'node:fs/promises'
@@ -157,7 +157,48 @@ function patchIhdr(png: Buffer, width: number, height: number, colorType?: numbe
   return result
 }
 
-test('loadImage rejects malformed raster images without terminating the process (issue #1309)', (t) => {
+function runInChildProcess(
+  script: string,
+  args: string[],
+): Promise<{
+  status: number | null
+  signal: NodeJS.Signals | null
+  stdout: string
+  stderr: string
+}> {
+  return new Promise((resolve, reject) => {
+    const child = spawn(process.execPath, ['-e', script, ...args], {
+      stdio: ['ignore', 'pipe', 'pipe'],
+    })
+    let stdout = ''
+    let stderr = ''
+
+    child.stdout.setEncoding('utf8')
+    child.stdout.on('data', (chunk) => {
+      stdout += chunk
+    })
+    child.stderr.setEncoding('utf8')
+    child.stderr.on('data', (chunk) => {
+      stderr += chunk
+    })
+
+    const timer = setTimeout(() => {
+      child.kill()
+      reject(new Error('Child process timed out after 10 seconds'))
+    }, 10_000)
+
+    child.on('error', (error) => {
+      clearTimeout(timer)
+      reject(error)
+    })
+    child.on('close', (status, signal) => {
+      clearTimeout(timer)
+      resolve({ status, signal, stdout, stderr })
+    })
+  })
+}
+
+test('loadImage rejects malformed raster images without terminating the process (issue #1309)', async (t) => {
   const canvas = createCanvas(64, 64)
   const ctx = canvas.getContext('2d')
   ctx.fillStyle = '#345678'
@@ -193,12 +234,8 @@ test('loadImage rejects malformed raster images without terminating the process 
   `
 
   for (const [name, image] of malformedImages) {
-    const child = spawnSync(process.execPath, ['-e', script, entrypoint, image.toString('base64')], {
-      encoding: 'utf8',
-      timeout: 10_000,
-    })
+    const child = await runInChildProcess(script, [entrypoint, image.toString('base64')])
 
-    t.falsy(child.error, `${name}: ${child.error?.message}`)
     t.is(child.signal, null, `${name}: child terminated with ${child.signal}\n${child.stderr}`)
     t.is(child.status, 0, `${name}: child exited with ${child.status}\n${child.stderr}`)
     t.regex(child.stdout, /^rejected: Unsupported image type\n$/, name)

--- a/__test__/loadimage.spec.ts
+++ b/__test__/loadimage.spec.ts
@@ -1,3 +1,4 @@
+import { spawnSync } from 'node:child_process'
 import { join, dirname } from 'node:path'
 import fs from 'node:fs'
 import { readFile } from 'node:fs/promises'
@@ -130,4 +131,86 @@ test('loadImage settles on tiny invalid Buffer (issue #1255)', async (t) => {
 test('loadImage settles on invalid base64 data URL (issue #1255)', async (t) => {
   const result = await settles(loadImage('data:image/png;base64,=').catch((e) => e))
   t.not(result, TIMEOUT_SENTINEL, 'loadImage("data:image/png;base64,=") hung')
+})
+
+// Regression tests for https://github.com/Brooooooklyn/canvas/issues/1309
+// Run malformed inputs in child processes so a native crash is reported as a
+// test failure instead of terminating the entire test runner.
+
+function crc32(buffer: Buffer): number {
+  let crc = -1
+  for (const byte of buffer) {
+    crc ^= byte
+    for (let bit = 0; bit < 8; bit++) {
+      crc = (crc >>> 1) ^ (crc & 1 ? 0xedb88320 : 0)
+    }
+  }
+  return (crc ^ -1) >>> 0
+}
+
+function patchIhdr(png: Buffer, width: number, height: number, colorType?: number): Buffer {
+  const result = Buffer.from(png)
+  result.writeUInt32BE(width, 16)
+  result.writeUInt32BE(height, 20)
+  if (colorType !== undefined) result[25] = colorType
+  result.writeUInt32BE(crc32(result.subarray(12, 29)), 29)
+  return result
+}
+
+test('loadImage rejects malformed raster images without terminating the process (issue #1309)', (t) => {
+  const canvas = createCanvas(64, 64)
+  const ctx = canvas.getContext('2d')
+  ctx.fillStyle = '#345678'
+  ctx.fillRect(0, 0, 64, 64)
+  const png = canvas.toBuffer('image/png')
+
+  const jpegSignature = Buffer.from(png)
+  jpegSignature.set([0xff, 0xd8, 0xff, 0xe0], 0)
+
+  const malformedImages = [
+    ['PNG truncated mid-IDAT', png.subarray(0, 40)],
+    ['PNG truncated after IHDR', png.subarray(0, 33)],
+    [
+      'PNG signature with random data and IEND',
+      Buffer.concat([png.subarray(0, 8), Buffer.from('malformed raster payload'), png.subarray(-12)]),
+    ],
+    ['PNG with invalid color type', patchIhdr(png, 64, 64, 9)],
+    ['PNG with zero dimensions', patchIhdr(png, 0, 0)],
+    ['PNG payload with a JPEG signature', jpegSignature],
+    ['PNG with excessive dimensions', patchIhdr(png, 100_000, 100_000)],
+  ] as const
+
+  const entrypoint = join(__dirname, '..', 'index.js')
+  const script = `
+    const { loadImage } = require(process.argv[1])
+    loadImage(Buffer.from(process.argv[2], 'base64')).then(
+      () => {
+        console.error('unexpectedly resolved')
+        process.exitCode = 2
+      },
+      (error) => console.log('rejected:', error.message),
+    )
+  `
+
+  for (const [name, image] of malformedImages) {
+    const child = spawnSync(process.execPath, ['-e', script, entrypoint, image.toString('base64')], {
+      encoding: 'utf8',
+      timeout: 10_000,
+    })
+
+    t.falsy(child.error, `${name}: ${child.error?.message}`)
+    t.is(child.signal, null, `${name}: child terminated with ${child.signal}\n${child.stderr}`)
+    t.is(child.status, 0, `${name}: child exited with ${child.status}\n${child.stderr}`)
+    t.regex(child.stdout, /^rejected: Unsupported image type\n$/, name)
+  }
+})
+
+test('loadImage preserves Skia partial decoding for sufficiently complete PNG data (issue #1309)', async (t) => {
+  const canvas = createCanvas(64, 64)
+  const png = canvas.toBuffer('image/png')
+
+  const image = await loadImage(png.subarray(0, 119))
+
+  t.is(image.width, 64)
+  t.is(image.height, 64)
 })

--- a/package.json
+++ b/package.json
@@ -81,6 +81,7 @@
     "@jimp/png": "^0.22.12",
     "@napi-rs/cli": "^3.5.0",
     "@napi-rs/webcodecs": "^1.1.1",
+    "@node-rs/crc32": "^1.10.6",
     "@octokit/rest": "^22.0.1",
     "@oxc-node/cli": "^0.1.0",
     "@oxc-node/core": "^0.1.0",

--- a/skia-c/skia_c.cpp
+++ b/skia-c/skia_c.cpp
@@ -1,6 +1,7 @@
 #include <assert.h>
 #include <math.h>
 #include <algorithm>
+#include <memory>
 #include <optional>
 #include <vector>
 
@@ -1976,11 +1977,14 @@ void skiac_sk_data_destroy(skiac_data* c_data) {
 
 // Bitmap
 
-void skiac_bitmap_make_from_buffer(const uint8_t* ptr,
+bool skiac_bitmap_make_from_buffer(const uint8_t* ptr,
                                    size_t size,
                                    skiac_bitmap_info* bitmap_info) {
   auto data = SkData::MakeWithoutCopy(reinterpret_cast<const void*>(ptr), size);
   auto codec = SkCodec::MakeFromData(data);
+  if (!codec) {
+    return false;
+  }
   auto info = codec->getInfo();
   // PNG/WebP/GIF codecs report unpremultiplied alpha (required by those
   // formats on disk). Ask the codec to output premultiplied pixels so that
@@ -1991,9 +1995,15 @@ void skiac_bitmap_make_from_buffer(const uint8_t* ptr,
     info = info.makeAlphaType(kPremul_SkAlphaType);
   }
   auto row_bytes = info.minRowBytes();
-  auto bitmap = new SkBitmap();
-  bitmap->allocPixels(info);
-  codec->getPixels(info, bitmap->getPixels(), row_bytes);
+  auto bitmap = std::make_unique<SkBitmap>();
+  if (!bitmap->tryAllocPixels(info)) {
+    return false;
+  }
+  auto result = codec->getPixels(info, bitmap->getPixels(), row_bytes);
+  if (result != SkCodec::kSuccess && result != SkCodec::kIncompleteInput &&
+      result != SkCodec::kErrorInInput) {
+    return false;
+  }
   auto dimension = codec->dimensions();
   auto origin = codec->getOrigin();
   auto width = dimension.width();
@@ -2008,25 +2018,30 @@ void skiac_bitmap_make_from_buffer(const uint8_t* ptr,
       width = height;
       height = dimension.width();
     }
-    auto oriented_bitmap = new SkBitmap();
+    auto oriented_bitmap = std::make_unique<SkBitmap>();
     auto oriented_bitmap_info =
         SkImageInfo::Make(width, height, info.colorType(), info.alphaType());
-    oriented_bitmap->allocPixels(oriented_bitmap_info);
-    auto canvas = new SkCanvas(*oriented_bitmap);
+    if (!oriented_bitmap->tryAllocPixels(oriented_bitmap_info)) {
+      return false;
+    }
+    SkCanvas canvas(*oriented_bitmap);
     auto matrix = SkEncodedOriginToMatrix(origin, width, height);
-    canvas->setMatrix(matrix);
+    canvas.setMatrix(matrix);
     auto image = SkImages::RasterFromBitmap(*bitmap);
-    canvas->drawImage(image, 0, 0);
-    delete canvas;
+    if (!image) {
+      return false;
+    }
+    canvas.drawImage(image, 0, 0);
     oriented_bitmap->setImmutable();
-    bitmap_info->bitmap = reinterpret_cast<skiac_bitmap*>(oriented_bitmap);
-    delete bitmap;
+    bitmap_info->bitmap =
+        reinterpret_cast<skiac_bitmap*>(oriented_bitmap.release());
   } else {
     bitmap->setImmutable();
-    bitmap_info->bitmap = reinterpret_cast<skiac_bitmap*>(bitmap);
+    bitmap_info->bitmap = reinterpret_cast<skiac_bitmap*>(bitmap.release());
   }
   bitmap_info->width = width;
   bitmap_info->height = height;
+  return true;
 }
 
 bool skiac_bitmap_make_from_svg(const uint8_t* data,

--- a/skia-c/skia_c.hpp
+++ b/skia-c/skia_c.hpp
@@ -1183,7 +1183,7 @@ void skiac_canvas_draw_sk_image(skiac_canvas* c_canvas,
 void skiac_sk_data_destroy(skiac_data* c_data);
 
 // Bitmap
-void skiac_bitmap_make_from_buffer(const uint8_t* ptr,
+bool skiac_bitmap_make_from_buffer(const uint8_t* ptr,
                                    size_t size,
                                    skiac_bitmap_info* bitmap_info);
 bool skiac_bitmap_make_from_svg(const uint8_t* data,

--- a/src/image.rs
+++ b/src/image.rs
@@ -1024,11 +1024,17 @@ impl<'env> ScopedTask<'env> for BitmapDecoder {
           .map_err(|e| Error::new(Status::InvalidArg, format!("Decode data url failed {e}")))?;
         if let Some(kind) = infer::get(&image_binary) {
           if kind.matcher_type() == infer::MatcherType::Image {
-            DecodeStatus::Ok(BitmapInfo {
-              data: Bitmap::from_buffer(image_binary.as_ptr().cast_mut(), image_binary.len()),
-              is_svg: false,
-              decoded_image: None,
-            })
+            if let Some(bitmap) =
+              Bitmap::from_buffer(image_binary.as_ptr().cast_mut(), image_binary.len())
+            {
+              DecodeStatus::Ok(BitmapInfo {
+                data: bitmap,
+                is_svg: false,
+                decoded_image: None,
+              })
+            } else {
+              DecodeStatus::InvalidImage
+            }
           } else {
             DecodeStatus::InvalidImage
           }
@@ -1063,11 +1069,15 @@ impl<'env> ScopedTask<'env> for BitmapDecoder {
       false
     } {
       // Other image formats detected by infer (PNG, JPEG, GIF, WebP, etc.)
-      DecodeStatus::Ok(BitmapInfo {
-        data: Bitmap::from_buffer(data_ref.as_ptr().cast_mut(), length),
-        is_svg: false,
-        decoded_image: None,
-      })
+      if let Some(bitmap) = Bitmap::from_buffer(data_ref.as_ptr().cast_mut(), length) {
+        DecodeStatus::Ok(BitmapInfo {
+          data: bitmap,
+          is_svg: false,
+          decoded_image: None,
+        })
+      } else {
+        DecodeStatus::InvalidImage
+      }
     } else if is_svg_image(&data_ref, length) {
       let font = get_font().map_err(SkError::from)?;
       if (self.width - -1.0).abs() > f64::EPSILON && (self.height - -1.0).abs() > f64::EPSILON {

--- a/src/sk.rs
+++ b/src/sk.rs
@@ -1008,7 +1008,11 @@ pub mod ffi {
 
     pub fn skiac_sk_data_destroy(c_data: *mut skiac_data);
 
-    pub fn skiac_bitmap_make_from_buffer(ptr: *mut u8, size: usize, info: *mut skiac_bitmap_info);
+    pub fn skiac_bitmap_make_from_buffer(
+      ptr: *mut u8,
+      size: usize,
+      info: *mut skiac_bitmap_info,
+    ) -> bool;
 
     pub fn skiac_bitmap_make_from_svg(
       data: *const u8,
@@ -4129,7 +4133,7 @@ unsafe impl Send for Bitmap {}
 unsafe impl Sync for Bitmap {}
 
 impl Bitmap {
-  pub fn from_buffer(ptr: *mut u8, size: usize) -> Self {
+  pub fn from_buffer(ptr: *mut u8, size: usize) -> Option<Self> {
     let mut bitmap_info = ffi::skiac_bitmap_info {
       bitmap: ptr::null_mut(),
       width: 0,
@@ -4137,8 +4141,11 @@ impl Bitmap {
       is_canvas: false,
     };
     unsafe {
-      ffi::skiac_bitmap_make_from_buffer(ptr, size, &mut bitmap_info);
-      Bitmap(bitmap_info)
+      if ffi::skiac_bitmap_make_from_buffer(ptr, size, &mut bitmap_info) {
+        Some(Bitmap(bitmap_info))
+      } else {
+        None
+      }
     }
   }
 

